### ISSUE TITLE
the merge method from AnnotationsBag is broken and need a fix

### ISSUE THREAD

**Issue by nyamsprod** [marcioAlmada/annotations#13]
**tl;dr**:
If you merge 2 AnnotationsBag object with different parserRules, the resulting AnnotationsBag will behave incorrectly. the `export` method will export data that can not be reach using the  `get` or the `getAsArray` method

**Usually I would directly submit a fix using a pull request ... but I'm too busy right now and It may take me 2 weeks to do it properly with the unit testing and all so I'll include here 2 proposals to fix the bug since I think this is urgent.**
## The problem

Imagine you have 2 AnnotationsBag (`$lowerBag` and `$upperBag`) with 2 differents PaserRules.

`$lowerBag` ParserRules states that arguments must be lowercase
`$upperBag` ParserRules states that arguments must be uppercase

``` php
$lowerBag->export(); // ['toto' => true, 'tata' => true, 'foo' => 'bar'];
$upperBag->export(); // ['TOTO' => true, 'TATA' => true, 'FOO' => 'bar'];
$lowerBag->merge($upperBag);
$lowerBag->export(); // ['toto' => true, 'tata' => true, 'foo' => 'bar', 'TOTO' => true, 'TATA' => true, 'FOO' => 'bar'];
$lowerBag->get('TOTO'); // will throw an InvalidArgumentException !!!
```

`$lowerBag` behaviour following the `merge` call is totally buggy and must be fix!!
### Proposal 1: keep `AnnotationsBag` immutable

Making the `merge` method behave like `grep` and `useNamespace` to keep `AnnotationsBag` immutable.

``` php
public function merge(AnnotationsBag $bag)
{
    $res = array_merge($this->attributes, $bag->export());
    return new static($res, $this->rules);
}
```

This fix the problem.

**Benefits:**
- The resulting `AnnotationsBag` filter out the wrong values on constructor call.
- AnnotationsBag stay immutable

**Disadvantages:**
- The  `merge` method will no longer affect the parent object like intended (IMHO this is not a disadvantage)
### Proposal 2: add a `replace` method to `AnnotationsBag`

Create a new method called `replace` (I'll leave to you to decide this method visibility) making it public would open a whole new range of possibilities :) .

``` php
public function replace(array $attributes = array())
{
    foreach ($attributes as $key => $value) {
        if ($this->rules->isKeyValid($key)) {
            $this->attributes[$key] = $attributes[$key];
        }
    }

    return $this;
}
```

With this function you can then update the `__construct` and `merge` methods as follow:

``` php
public function __construct(array $attributes, ParserRulesInterface $rules)
{
    $this->rules = $rules;
    $this->replace($attributes);
}

public function merge(AnnotationsBag $bag)
{
    return $this->replace($bar->export());
}
```

**Benefits:**
- you get a new method `replace`
- you don't change `merge` behaviour (IMHO this **is** a disadvantage)
## Final notes

Keep in mind that those 2 proposals are not exclusive. You can still implement the `replace` method even if you choose to use proposal 1.

I hope this will help you fix the bug

nyamsprod
